Objects/TI4_Helpers/TI4_FactionHelper.ttslua

```diff
@@ -1122,6 +1126,703 @@ function tintTokens(params)
+    if owningObject.tag == 'Card' then
+        purgeCard(owningObject)
+    end
+
+    -- Place frontier tokens on systems that need one
+    _exploreHelper.placeFrontierTokens()
+
+    -- Wait for frontier tokens to land before exploring them.
+    local waitUntil = Time.time + 3
+    while Time.time < waitUntil do
+        coroutine.yield(0)
+    end
+
+    --Find faction owning this hero ability (and it's color)
+    local factionTokenName = false
+    local factionColor = false
+    for color, faction in pairs(allFactions(false)) do
+        if faction.hero == owningObject.getName() then
+            factionTokenName = faction.tokenName
+            factionColor = color
+            break
+        end
+    end
+    assert(factionTokenName and factionColor, 'Multiverse Shift: Placed Frontier Tokens, but cannot explore them with missing owning faction.')
+
+    --Map frontier tokens to system they're sitting on
+    local frontierTokens = _exploreHelper.getAllFrontierTokens()
+    local frontierTokenToPosition = {}
+    for _, token in ipairs(frontierTokens) do
+        frontierTokenToPosition[token] = token.getPosition()
+    end
+
+    local frontierTokenToSystem = _systemHelper.systemsFromPositions(frontierTokenToPosition)
+
+    --Find all units
+    local allUnits = _unitHelper.getUnits()
+
+    --Filter down to owning faction ships
+    --NOTE: Can skip units lacking faction/color; they'll be alongside other units.
+    local factionShipToPosition = {}
+    for _, unit in ipairs(allUnits) do
+        if unit.unitType ~= 'Infantry' and unit.unitType ~= 'Mech' and unit.unitType ~= 'Space Dock' and unit.unitType ~= 'PDS' then --ships only
+            if unit.factionTokenName and unit.factionTokenName == factionTokenName then
+                factionShipToPosition[unit.guid] = unit.position
+            elseif unit.color and unit.color == factionColor then
+                factionShipToPosition[unit.guid] = unit.position
+            end
+        end
+    end
+
+    --Find each system containing owning faction ship
+    local shipToSystem = _systemHelper.systemsFromPositions(factionShipToPosition)
+    local systemsWithFactionShipsSet = {}
+    for ship, system in pairs(shipToSystem) do
+        if system then
+            systemsWithFactionShipsSet[system.guid] = true
+        end
+    end
+
+    --For each system containing a frontier token AND owning faction ship, highlight the system tile
+    local explorableTokensToSystem = {}
+    local systemsToExplore = false
+    for token, system in pairs(frontierTokenToSystem) do
+        if system and systemsWithFactionShipsSet[system.guid] then
+            explorableTokensToSystem[token] = system
+            systemsToExplore = true
+        end
+    end
+
+    -- While there are still systems to explore, blink those system tiles with the player color (2 minute timeout)
+    local timeout = Time.time + 120
+    while Time.time < timeout and systemsToExplore do
```

Comment:
So, in the pickup/delete handlers, checking the object's GUID against a list is pretty safe from a perf/stability perspective?

Event handlers will be another area where it's not obvious to me what the pitfalls are

Reply:
yeah, it would be bad if hundreds of objects had global event handlers but a dozen is ok.  They are also very fast, just removing an entry from a table for the most part.